pkg/services/cloudmigration/cloudmigrationimpl/cloudmigration.go

```diff
@@ -419,168 +425,163 @@ func (s *Service) RunMigration(ctx context.Context, uid string) (*cloudmigration
-	dashs, err := s.dashboardService.GetAllDashboards(ctx)
+	// create new directory for snapshot writing
+	snapshotUid := util.GenerateShortUID()
+	dir := filepath.Join("cloudmigration.snapshots", fmt.Sprintf("snapshot-%s-%s", snapshotUid, initResp.GMSSnapshotUID))
+	err = os.MkdirAll(dir, 0750)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("creating snapshot directory: %w", err)
 	}
 
-	var result []dashboards.Dashboard
-	for _, dashboard := range dashs {
-		result = append(result, *dashboard)
+	// save snapshot to the db
+	snapshot := cloudmigration.CloudMigrationSnapshot{
+		UID:            snapshotUid,
+		SessionUID:     sessionUid,
+		Status:         cloudmigration.SnapshotStatusInitializing,
+		EncryptionKey:  initResp.EncryptionKey,
+		UploadURL:      initResp.UploadURL,
+		GMSSnapshotUID: initResp.GMSSnapshotUID,
+		LocalDir:       dir,
 	}
-	return result, nil
-}
 
-func (s *Service) createMigrationRun(ctx context.Context, cmr cloudmigration.CloudMigrationSnapshot) (string, error) {
-	uid, err := s.store.CreateMigrationRun(ctx, cmr)
+	uid, err := s.store.CreateSnapshot(ctx, snapshot)
 	if err != nil {
-		s.log.Error("Failed to save migration run", "err", err)
-		return "", err
+		return nil, fmt.Errorf("saving snapshot: %w", err)
 	}
-	return uid, nil
+	snapshot.UID = uid
+
+	// start building the snapshot asynchronously while we return a success response to the client
+	go s.buildSnapshot(context.Background(), snapshot)
+
+	return &snapshot, nil
 }
 
-func (s *Service) GetMigrationStatus(ctx context.Context, runUID string) (*cloudmigration.CloudMigrationSnapshot, error) {
-	cmr, err := s.store.GetMigrationStatus(ctx, runUID)
+// GetSnapshot returns the on-prem version of a snapshot, supplemented with processing status from GMS
+func (s *Service) GetSnapshot(ctx context.Context, sessionUid string, snapshotUid string) (*cloudmigration.CloudMigrationSnapshot, error) {
+	ctx, span := s.tracer.Start(ctx, "CloudMigrationService.GetSnapshot")
+	defer span.End()
+
+	snapshot, err := s.store.GetSnapshotByUID(ctx, snapshotUid)
 	if err != nil {
-		return nil, fmt.Errorf("retrieving migration status from db: %w", err)
+		return nil, fmt.Errorf("fetching snapshot for uid %s: %w", snapshotUid, err)
 	}
-	return cmr, nil
-}
 
-func (s *Service) GetMigrationRunList(ctx context.Context, migUID string) (*cloudmigration.SnapshotList, error) {
-	runs, err := s.store.GetMigrationStatusList(ctx, migUID)
+	session, err := s.store.GetMigrationSessionByUID(ctx, sessionUid)
 	if err != nil {
-		return nil, fmt.Errorf("retrieving migration statuses from db: %w", err)
+		return nil, fmt.Errorf("fetching session for uid %s: %w", sessionUid, err)
 	}
 
-	runList := &cloudmigration.SnapshotList{Runs: []cloudmigration.MigrateDataResponseList{}}
-	for _, s := range runs {
-		runList.Runs = append(runList.Runs, cloudmigration.MigrateDataResponseList{
-			RunUID: s.UID,
-		})
+	if snapshot.Status == cloudmigration.SnapshotStatusPendingProcessing ||
```

Comment:
Could we make the conditions well named  method for readability? Maybe there's a concept here

Reply:
done